packages/caver-transaction/src/transactionTypes/abstractTransaction.js

```diff
@@ -0,0 +1,343 @@
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The caver-js library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+const _ = require('lodash')
+const Bytes = require('eth-lib/lib/bytes')
+const RLP = require('eth-lib/lib/rlp')
+const Hash = require('eth-lib/lib/hash')
+const TransactionHasher = require('../transactionHasher/transactionHasher')
+const utils = require('../../../caver-utils')
+const Keyring = require('../../../caver-wallet/src/keyring/keyring')
+const {
+    TX_TYPE_STRING,
+    TX_TYPE_TAG,
+    refineSignatures,
+    typeDetectionFromRLPEncoding,
+    isValidNumber,
+} = require('../transactionHelper/transactionHelper')
+const { KEY_ROLE } = require('../../../caver-wallet/src/keyring/keyringHelper')
+const { validateParams } = require('../../../caver-core-helpers/src/validateFunction')
+
+/**
+ * Abstract class that implements common logic for each transaction type.
+ * @class
+ */
+class AbstractTransaction {
+    /**
+     * Abstract class that implements common logic for each transaction type.
+     * In this constructor, type, tag, nonce, gasPrice, chainId, gas and signatures are set as transaction member variables.
+     *
+     * @constructor
+     * @param {string} typeString - The type string of transaction.
+     * @param {object} createTxObj - The parameters to create a transaction instance.
+     */
+    constructor(typeString, createTxObj) {
+        this.type = typeString
+        this.tag = TX_TYPE_TAG[typeString]
+
+        createTxObj.type = typeString
+
+        const err = validateParams(createTxObj)
+        if (err) throw err
+
+        this.gas = createTxObj.gas
+
+        // The variables below are values that the user does not need to pass to the parameter.
+        if (createTxObj.nonce !== undefined) this.nonce = createTxObj.nonce
+        if (createTxObj.gasPrice !== undefined) this.gasPrice = createTxObj.gasPrice
+        if (createTxObj.chainId !== undefined) this.chainId = createTxObj.chainId
+
+        this.signatures = createTxObj.signatures || []
+    }
+
+    /**
+     * @type {string}
+     */
+    get type() {
+        return this._type
+    }
+
+    set type(t) {
+        if (TX_TYPE_STRING[t] === undefined) throw new Error(`Invalid transaction type ${t}. Please refer 'caver.transaction.type'.`)
+        this._type = t
+    }
+
+    /**
+     * @type {string}
+     */
+    get tag() {
+        return this._tag
+    }
+
+    set tag(t) {
+        if (TX_TYPE_TAG[t] === undefined) throw new Error(`Invalid transaction tag ${t}. Please refer 'caver.transaction.tag'.`)
```

Comment:
Is it possible that we can have different values in `type` and `tag`?
If we change type, I think tag also should be updated. And, vice versa.

Reply:
Or, this is another option that we can remove `tag` setter.

Reply:
I removed tag variable. :) 
I think user cannot set type string also, i removed type setter.